### PR TITLE
[FIRRTL][HW] Shared InstancePathCache between HW and FIRRTL

### PR DESCRIPTION
**Pull request by nandor**
Instance path enumeration considers public modules as starting points of the enumeration. Depends on #3833 which changes the visibility of internal modules to allow the root to be identifies as the sole public module across both `hw` and `firrtl`.

### CODE REVIEW

File: test/Dialect/FIRRTL/emit-metadata.mlir [hunk 1]
```diff
@@ -125,14 +125,14 @@ firrtl.circuit "top"
       firrtl.instance dut @dut()
       firrtl.instance mem1 @Mem1()
     }
-    firrtl.module @Mem1() {
+    firrtl.module private @Mem1() {
```
Comment (fabianschuiki) [inline test/Dialect/FIRRTL/emit-metadata.mlir:128]: Just out of curiosity: do we have to make all these modules private for the tests to work? Don't they just default to public which for the majority of tests is good enough?

**Review by fabianschuiki — COMMENTED** [review]
Love this refactoring!

File: test/Dialect/FIRRTL/emit-metadata.mlir (hunk 1 above)
Comment (nandor) [reply test/Dialect/FIRRTL/emit-metadata.mlir:128]: Yes, we assume that there is only one public module originating from the firrtl toplevel.

File: test/Dialect/FIRRTL/emit-omir-errors.mlir [hunk 2]
```diff
@@ -8,64 +8,10 @@ firrtl.circuit "Top" attributes {annotations = [{
...
-// expected-error @+3 {{tracked OMIR target of type `OMDontTouchedReferenceTarget` was deleted}}
-// expected-note @+2 {{`OMDontTouchedReferenceTarget` should never be deleted}}
-// expected-note @+1 {{original path: `~Foo|Bar`}}
-firrtl.circuit "Foo" attributes {annotations = [{
-  class = "freechips.rocketchip.objectmodel.OMIRAnnotation",
-  nodes = [{info = #loc, id = "OMID:0", fields = {
-    a = {info = #loc, index = 1, value = {omir.tracker, id = 0, path = "~Foo|Bar", type = "OMDontTouchedReferenceTarget"}}
-  }}]
-}]} {
-  firrtl.module @Foo() {}
```
[the remark this answers is not in the document]
Comment (nandor) [reply test/Dialect/FIRRTL/emit-omir-errors.mlir:70]: :facepalm:

**Pull request closed without merging**